File: Mobiles/android/app/src/main/java/com/grafana/quickpizza/features/auth/domain/AuthRepository.kt
package com.grafana.quickpizza.features.auth.domain

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.grafana.quickpizza.core.api.ApiClient
import com.grafana.quickpizza.core.storage.TokenStorage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepository @Inject constructor(
    private val apiClient: ApiClient,
    private val tokenStorage: TokenStorage,
) {
    val isAuthenticated: Boolean
        get() = tokenStorage.token != null

    val username: String?
        get() = tokenStorage.username

    suspend fun login(username: String, password: String) = withContext(Dispatchers.IO) {
        val response = apiClient.post(
            "/api/users/token/login",
            body = LoginRequest(username, password),
            includeAuth = false,
        )
        if (!response.isSuccessful) {
            error("Login failed: HTTP ${response.code}")
        }
        val body = response.body?.string() ?: error("Empty response body")
        val loginResponse = Gson().fromJson(body, LoginResponse::class.java)
        tokenStorage.token = loginResponse.token
        tokenStorage.username = username
    }

    fun logout() {
        tokenStorage.clear()
    }

    private data class LoginRequest(
        val username: String,
        val password: String,
    )

    private data class LoginResponse(
        @SerializedName("token") val token: String,
    )
}

File: Mobiles/react-native/src/features/auth/domain/authRepository.ts
import { apiPost } from '../../../core/api/apiClient';
import * as tokenStorage from '../../../core/storage/tokenStorage';

export async function login(username: string, password: string): Promise<boolean> {
  try {
    const response = await apiPost(
      '/api/users/token/login',
      { username, password },
      { includeAuth: false },
    );

    if (response.ok) {
      const json = (await response.json()) as { token?: string };
      const token = json.token;
      if (token) {
        await tokenStorage.saveSession(token, username);
        return true;
      }
    }
    return false;
  } catch {
    return false;
  }
}

export async function logout(): Promise<void> {
  await tokenStorage.clearSession();
}

export async function loadSession(): Promise<tokenStorage.StoredSession> {
  return tokenStorage.loadSession();
}
